Implement Move To Trash feature

oneeyeman1 opened the pull request:
This is a second attempt to push "Move to Trash" feature to the library.

I made this function static and I hope I fixed all the documentation things.

Please review.

Next will be an asynchronous implementation of this feature.

Thank you and sorry for such a long delay.

oneeyeman1 committed:
Implement the feature - WIP

oneeyeman1 committed:
Implement the feature - WIP

a-wi committed:
Execute GUI tests under wxOSX

a-wi committed:
Skip tests that currently fail under wxMac

oneeyeman1 committed:
Modify build system files

oneeyeman1 committed:
Modify build system files

vadz committed:
Remove unnecessary local variable in wxComboCtrlBase code

There doesn't seem to be any point in using local "customBorder" instead
of member "m_widthCustomBorder", so remove the unnecessary local
variable and use the member variable instead.

No real changes.

vadz committed:
Allow using custom main window in wxComboCtrl

Add wxComboCtrl::SetMainControl() which can be used to use some other
window instead of the default wxTextCtrl as the main window of the
combo control.

Update the sample and the documentation to show the new function.

iwbnwif committed:
Switch to using GSpell for wxTextCtrl proof checker in wxGTK

Use GSpell rather than GtkSpell because GSpell also allows to check
single line text controls.

Closes https://github.com/wxWidgets/wxWidgets/pull/2516

oneeyeman1 committed:
Generate makefiles and fix GTK build

oneeyeman1 committed:
Generate makefiles and fix GTK build

oneeyeman1 committed:
Fix OSX build

oneeyeman1 committed:
Fix OSX build

oneeyeman1 committed:
Fix spacing error

oneeyeman1 committed:
Fix spacing error

oneeyeman1 committed:
Fix cross-compile errors

oneeyeman1 committed:
Fix cross-compile errors

oneeyeman1 committed:
Merge branch 'ticket_9446_part2' of github.com:oneeyeman1/wxWidgets into ticket_9446_part2

oneeyeman1 committed:
Trying to fix iphone build

oneeyeman1 committed:
Trying to fix iphone build

PBfordev commented on include/wx/desktopenv.h:
```diff
@@ -0,0 +1,33 @@
+// Author:      Igor Korot
+// Created:     2021-09-05
+// Copyright:   (c) 2021 wxWidgets development team
+// Licence:     wxWidgets licence
```
```suggestion
// Licence:     wxWindows licence
```

Applies everywhere.

oneeyeman1 replied:
@PBfordev ,
I fixed those, however, there are 2 more files with mentioning of "wxWidgets license" - mimetype.h/cpp.

I don't know what the plan is for them to be.

Vadim?

PBfordev commented on include/wx/msw/desktopenv.h:
```diff
@@ -0,0 +1,24 @@
+{
+public:
+    static bool MoveToRecycleBin(const wxString &path);
+protected:
```
Remove the unused access specifier.

PBfordev commented on interface/wx/desktopenv.h:
```diff
@@ -0,0 +1,29 @@
+or Recycle Bin. The call to this function is synchronous, which means that
+application will wait for the move to finish.
+
+@since 3.1.5
```
```suggestion
@since 3.1.6
```

PBfordev commented on interface/wx/desktopenv.h:
```diff
@@ -0,0 +1,29 @@
+application will wait for the move to finish.
+
+@since 3.1.5
+This conversation was marked as resolved by oneeyeman1
```
```suggestion
```

PBfordev commented on interface/wx/desktopenv.h:
```diff
@@ -0,0 +1,29 @@
+{
+public:
+    /**
+    Moves the file/folder to the Recycle Bin/Trash. The file name should be an absolute path
```
```suggestion
    Moves the file/folder to the Recycle Bin/Trash. The path should be absolute.
```

The implementation makes the path absolute.

PBfordev commented on src/common/desktopenvcmn.cpp:
```diff
@@ -0,0 +1,43 @@
+// for compilers that support precompilation, includes "wx.h".
+#include "wx/wxprec.h"
+
+#ifdef __BORLANDC__
```
BCC is not supported anymore, its stuff needs to be removed everywhere.

PBfordev commented on src/common/desktopenvcmn.cpp:
```diff
@@ -0,0 +1,43 @@
+#include "wx/desktopenv.h"
+
+/* static */
+bool wxDesktopEnvBase::MoveToRecycleBin(const wxString &path)
```
```suggestion
bool wxDesktopEnvBase::MoveToRecycleBin(const wxString& path)
```
Applies everywhere.

PBfordev commented on src/common/desktopenvcmn.cpp:
```diff
@@ -0,0 +1,43 @@
+    wxString tempPath = name.GetFullPath();
+    if( !wxFileExists( tempPath ) && !wxDirExists( tempPath ) )
+    {
+        wxLogSysError( _( "Failed to move '%s' to Recycle Bin" ), path.c_str() );
```
```suggestion
        wxLogSysError( _( "Failed to move '%s' to Recycle Bin" ), path);
```

Additionally, I am not sure if `wxLogSysError()` is the best fit here. Do we always have last error set here?

PBfordev commented on src/common/desktopenvcmn.cpp:
```diff
@@ -0,0 +1,43 @@
+{
+    bool result = true;
+    wxFileName name = wxFileName( path );
+    name.MakeAbsolute();
```
I would prefer separating variable declarations from the rest of the code by an empty line.

PBfordev commented on src/msw/desktopenv.cpp:
```diff
@@ -0,0 +1,54 @@
+        fileOp.fFlags = FOF_ALLOWUNDO | FOF_NOERRORUI | FOF_NOCONFIRMATION | FOF_SILENT;
+        res = SHFileOperation( &fileOp );
+        if( res != 0 )
+            wxLogSysError( _( "Failed to move '%s' to Recycle Bin" ), path );
```
`wxLogSysError()` should not be used here: `SHFileOperation()` does not affect the last error code.

Either just display the simple error message or if you want to display the specific error, you could do something like this
```cpp
wxLogError(_( "Moving '%s' to Recycle Bin failed with error 0x%08lx (%s)."),
    path, (long)res, wxSysErrorMsgStr(res));
```

PBfordev reviewed (commented):
I still do not understand the API, i.e., why have a class with a single, static, method. I guess this is supposed to be extended in the future. 

Even if so, why not submit it complete so we have a big picture available?

Nevertheless, I added a bunch of (mostly) nitpick comments.

PBfordev commented on tests/file/desktopenv.cpp:
```diff
@@ -0,0 +1,47 @@
+
+TEST_CASE_METHOD(DesktopEnvTestCase, "DesktopEnvTestCase::MoveToTrash")
+{
+#if defined(__WXMSW__) || defined(__WXGTK__) || ( defined(__WXOSX__) && !( wxOSX_USE_IPHONE ) ) )
```
```suggestion
#if defined(__WXMSW__) || defined(__WXGTK__) || ( defined(__WXOSX__) && !( wxOSX_USE_IPHONE ) )
```

PBfordev commented on interface/wx/desktopenv.h:
```diff
@@ -0,0 +1,29 @@
+    /**
+    Moves the file/folder to the Recycle Bin/Trash. The file name should be an absolute path
+
+    @returns true on success
```
I think return values are usually described in wxWidgets documentation like this
```suggestion
    @return @true if the file/folder was successfully moved to the Recycle Bin/Trash.
```

oneeyeman1 committed:
Fixing license name

oneeyeman1 committed:
Fixing license name

oneeyeman1 committed:
Remove unused keyword

oneeyeman1 committed:
Remove unused keyword

oneeyeman1 committed:
Fixing documentation

oneeyeman1 committed:
Fixing documentation

oneeyeman1 committed:
Remove Borland references

oneeyeman1 committed:
Remove Borland references

oneeyeman1 committed:
Fixes to the source code

oneeyeman1 committed:
Fixes to the source code

oneeyeman1 committed:
Fixing compilation of the tests

oneeyeman1 committed:
Fixing compilation of the tests

oneeyeman1 commented:
@PBfordev ,
Thank you for your input.
For the API - I plan to add asynchronous functions and possibly Restore From Trash/Recycle Bin.

And maybe I can add the wrapper for this on the global level - whatever Vadim will prefer.

Now - I'm also curious - where/how do I fix MSW cross-compile error? Add the missing symbol to msw/private.h?

oneeyeman1 committed:
Fix return statement in the docs

oneeyeman1 committed:
Fix return statement in the docs

oneeyeman1 committed:
Remove remnants of the code

oneeyeman1 committed:
Remove remnants of the code

vadz committed:
Fix crash due to missing retain in wxUILocale wxMac code

We need to retain [NSLocale availableLocaleIdentifiers] result too, as
it doesn't copy the returned object.

vadz committed:
Slightly simplify running the GUI tests in the Mac CI workflow

No real changes, just don't use pipefail option when we're not using
pipelines at all -- this was somewhat confusing.

vadz committed:
Merge branch 'combo-non-text'

Allow using non-text control as main window of wxComboCtrl.

See https://github.com/wxWidgets/wxWidgets/pull/2515

a-wi committed:
Fix Window::SizerErrors test in release build

In release build no exceptions are thrown by wxASSERT/wxCHECK
so there is no need to check them.

a-wi committed:
Execute Window::SizerErrors test under wxMac

oneeyeman1 commented:
And I have a question:

Would it make sense to have as a public function:

static bool FilesInTrahs(const wxVector<wxString> &);

?

TIA!!

vadz committed:
Fix handling multiple leading backslashes in wxFileName under MSW

Extend the existing workaround to work not only with explicitly
wxPATH_DOS paths, but with paths implicitly using the DOS format due to
it being the default format for the current platform.

Closes #19261.

vadz committed:
Fix wxPATH_RMDIR_FULL documentation

Explain what this flag really does in Rmdir() documentation itself, as
the old wording was confusing.

Closes #16682.

vadz committed:
Fix wxFileName::Mkdir("/foo") under MSW

Creating a directory with the leading path separator unexpectedly
created it under the current directory rather than in the root of the
current drive under MSW, due to the path being considered relative, in
spite of starting with the path separator, because it didn't have the
volume.

Fix this by avoiding the use of IsAbsolute() in Mkdir() and checking for
m_relative instead, as it seems the safest possible fix for this bug
because changing IsAbsolute() to return true in this case might change
the behaviour of the existing code.

Closes #4470.

vadz committed:
Replace old hack in wxFileName::Assign() with better solution

Replace the workaround introduced as a "temporary fix to avoid breaking
backwards compatibility in 2.8" back in 9e1c7236e0 (don't treat foo in
c:\\foo\bar as network share, 2006-12-17) with a better solution
proposed in the comment in that commit, i.e. don't even try to extract
the volume from the path if we already have the volume separately.

This commit is best viewed ignoring whitespace-only changes.

vadz committed:
Fix handling of single letter shares in UNC paths in wxFileName

This comes at the price of breaking compatibility and returning
"\\share" rather than just "share" from wxFileName::GetVolume() for the
UNC paths. This breakage seems justified because it is required in order
to allow application code to distinguish between paths "x:\foo" and
"\\x\foo", which was previously impossible as GetVolume() returned just
"x" in both cases.

Document this change, adjust the existing checks for the new GetVolume()
semantics and add a new test which passes now, but didn't pass before.

Closes #19255.

This commit is best viewed ignoring whitespace-only changes.

vadz committed:
Add another test case for wxPATH_NORM_DOTS

This was reported to not work but does work now.

See #15012.

vadz committed:
Fix wxFileName unit test under Unix after recent changes

Don't use GetVolumeSeparator() to combine UNC or GUID volumes with
paths, this doesn't work because this kind of paths doesn't contain
colons at all.

Update the documentation to mention this.

vadz commented on include/wx/desktopenv.h:
```diff
@@ -0,0 +1,33 @@
+    wxDECLARE_NO_COPY_CLASS(wxDesktopEnvBase);
+};
+
+#if defined (__WXMSW__)
+#include "wx/msw/desktopenv.h"
+#elif defined(__WXGTK20__)
+#include "wx/gtk/desktopenv.h"
+#elif defined(__WXOSX__)
+#include "wx/osx/desktopenv.h"
+#else
+#include "wx/generic/desktopenv.h"
+#endif
```
This should be just removed, as well as the corresponding header files themselves.

vadz commented on include/wx/desktopenv.h:
```diff
@@ -0,0 +1,33 @@
+
+#include "wx/defs.h"
+
+class WXDLLIMPEXP_CORE wxDesktopEnvBase
```
What's the point of having a "Base" class when there are no virtual functions here? This should be just `wxDesktopEnv` itself, it's a namespace-like class, it doesn't make any sense to inherit from it.

vadz commented on interface/wx/desktopenv.h:
```diff
@@ -0,0 +1,28 @@
+/**
+@class wxDesktopEnv
+
+wxDesktopEnv contains the function for moving the file/folder to Trash
+or Recycle Bin. The call to this function is synchronous, which means that
+application will wait for the move to finish.
```
There is no real need to mention that a function is sync, they all are by default.

vadz commented on src/common/desktopenvcmn.cpp:
```diff
@@ -0,0 +1,40 @@
+
+    name.MakeAbsolute();
+    wxString tempPath = name.GetFullPath();
+    if( !wxFileExists( tempPath ) && !wxDirExists( tempPath ) )
+    {
+        wxLogSysError( _( "Failed to move '%s' to Recycle Bin" ), path );
+        result = false;
+    }
```
Why should we bother checking this at all? If the file doesn't exist, moving it to trash should fail anyhow.

This entire function should be just removed.

oneeyeman1 replied:
Vadim,
Why even try if the file doesn't exist? It just doesn't make sense...
And this is exactly why the base class exist - to have this check in one place and don't have it in 3 places.

I may give it a try and see if it succeeds on all 3 supported platforms without crashing the program.

vadz commented on src/generic/desktopenv.cpp:
```diff
@@ -0,0 +1,31 @@
+
+#include "wx/desktopenv.h"
+
+bool wxDesktopEnv::MoveToRecycleBin(const wxString &WXUNUSED(path))
+{
+    wxFAIL_MSG( _( "Not implemented!" ) );
+    return true;
```
Failing here is probably not very useful. But it should return false, not true.

vadz commented on src/gtk/desktopenv.cpp:
```diff
@@ -0,0 +1,50 @@
+        GFile *f = g_file_new_for_path( path.c_str() );
+        gboolean res = g_file_trash( f, NULL, &error );
+        if( error && error->code == G_IO_ERROR_NOT_SUPPORTED )
+            wxLogSysError( _( "MoveToRecycleBin: Operation not supported." ) );
```
Why do you handle just this error code specifically? You should show the error message if any (i.e. if `error != NULL`) for any error code, use `wxGtkError` class from include/wx/gtk/private/error.h for it.

oneeyeman1 replied:
Because "Not supported" is different from any other failure, and the user should know about that specific one.

vadz replied:
Why is it different? Just give the associated error message, this should be clear enough and it should already say "not supported" in this case.

vadz commented on src/gtk/desktopenv.cpp:
```diff
@@ -0,0 +1,50 @@
+        else
+        {
+            result = true;
+            g_object_unref( f );
```
Does `g_file_trash()` take ownership of it on success? I.e. are you sure it doesn't need to be unref'd if it fails? I can't find documentation for this function but it would be surprising if it did this, so I think you need to use `wxGtkObject` to avoid leaks here.

vadz commented on src/osx/cocoa/desktopenv.mm:
```diff
@@ -0,0 +1,39 @@
+        BOOL result = [[NSFileManager defaultManager] trashItemAtURL:url resultingItemURL:nil error:nil];
+        if( result == NO )
+        {
+            wxLogSysError( _( "Failed to move '%s' to Recycle Bin" ), path );
```
Is it really called "recycle bin" under Mac?

vadz commented on tests/file/desktopenv.cpp:
```diff
@@ -0,0 +1,43 @@
+TEST_CASE_METHOD(DesktopEnvTestCase, "DesktopEnvTestCase::MoveToTrash")
+{
+#if defined(__WXMSW__) || defined(__WXGTK__) || ( defined(__WXOSX__) && !( wxOSX_USE_IPHONE ) )
+    std::ofstream out( "ffileinstream.test", std::ofstream::out );
+    wxString currentDir = wxGetCwd() + "/";
+    out << "test file to be moved to trash" << std::endl;
+    out.close();
```
Use `TestFile` from `tests/testfile.h` instead to ensure it's not left over.

vadz reviewed (changes requested):
This is still way too complicated and half of the changes don't make sense/are unnecessary. I don't know how to explain it, but you're adding a simple function, there is no need for any base classes, multiple platform-specific headers (with exactly the same contents!) or anything else.

The only interesting question for me is whether this should be a function in core or in base. As written right now, it could work without any GUI at all too, but this would require linking with gio under Linux. Not sure if it's worth it, but OTOH there is really nothing GUI-specific in this function.

vadz commented:
BTW, we already link with gio when using libsecret, which relies on it, so it probably wouldn't be a big deal to always link with it.

vadz committed:
Link only GUI libraries, not wxBase, with gspell

Don't add dependencies on gspell and all GUI libraries that it uses to
wxBase library, only link the GUI libraries with it.

See https://github.com/wxWidgets/wxWidgets/pull/2516

vadz committed:
Use PNG icons instead of BMP and XPM ones in the toolbar sample

Update the sample for the 21st century (with just a bit of delay) and
use PNGs under all platforms instead of using BMPs under MSW and XPMs
elsewhere.

Copy PNG icons from Tango, even if we already have almost of them in
art/tango because we plan to also use "2x" versions of them in this
sample soon.

vadz committed:
Get rid of INIT_TOOL_BMP() macro in the toolbar sample

It isn't really useful any more now that all ports use wxBITMAP_PNG()
and just obfuscates the code.

No real changes.

vadz committed:
Remove unnecessary calls to wxToolBar::SetToolBitmapSize()

This is unnecessary, just as an existing comment said, so don't bother
doing it.

vadz committed:
Remove extra blank lines from png2c.py output

Adding them just seems unnecessary.

vadz committed:
Show how to use higher resolution toolbar bitmaps in high DPI

Update the toolbar sample to show how to use bigger images for the
toolbar in high DPI.

oneeyeman1 committed:
Implement PR suggestions and add more functionality

oneeyeman1 committed:
Implement PR suggestions and add more functionality

oneeyeman1 committed:
Fix GTK build for previous commit

oneeyeman1 committed:
Fix GTK build for previous commit

oneeyeman1 committed:
Merge branch 'ticket_9446_part2' of github.com:oneeyeman1/wxWidgets into ticket_9446_part2

oneeyeman1 committed:
Fixing the test code

oneeyeman1 committed:
Fixing the test code

oneeyeman1 committed:
Merge branch 'ticket_9446_part2' of github.com:oneeyeman1/wxWidgets into ticket_9446_part2

oneeyeman1 committed:
Fixing the test code

oneeyeman1 committed:
Fixing the test code

oneeyeman1 committed:
Fixing the test code

oneeyeman1 committed:
Fixing the test code

vadz committed:
Restore recently removed #include wx/textctrl.h in wx/combobox.h

Some existing code relies on wxEVT_TEXT being defined after including
just wx/combobox.h too, so just keep this formally unnecessary but by
now impossible to remove #include removed in d3eafa4d18 (Fix compilation
of wxFormBuilder-generated code using wxTE_XXX, 2021-07-14).

See #14132.

Closes #19264.

a-wi committed:
Invalidate selection of item being removed from wxNSComboBoxControl

We need to explicitly invalidate selection of item being removed
because it's not done by NSComboBox (selection index is retained).

a-wi committed:
Retain selection of wxComboBox item whose text has been changed

In wxComboBox::SetString() implementation for wxMac the item is
first removed from the list so its selection is invalidated.
We have to retain it so we need to re-select just inserted item.

a-wi committed:
Execute ComboBoxTestCase under wxMac

vadz committed:
Update the shown file name in wxGTK wxFilePickerCtrl::SetPath()

This still doesn't work if the file doesn't exist, but at least shows
the correct file name if it does exist, which is better than nothing and
the best we can reasonably do when using the native control.

Closes #19163.

oneeyeman1 committed:
Fix MSW compilation

oneeyeman1 committed:
Fix MSW compilation

oneeyeman1 committed:
Trying to fix Cygwin build

oneeyeman1 committed:
Trying to fix Cygwin build

oneeyeman1 committed:
Trying to fix spacing

oneeyeman1 committed:
Trying to fix spacing

oneeyeman1 committed:
Merge branch 'ticket_9446_part2' of github.com:oneeyeman1/wxWidgets into ticket_9446_part2

a-wi committed:
Invalidate selection of item being removed from wxNSTableView

We need to explicitly invalidate selection of item being removed
because it's not done by NSTableView (selection index is retained).

a-wi committed:
Execute ListBox-related tests under wxMac

a-wi committed:
Adjust point to be tested in wxListBox::HitTest for macOS 11

On macOS >= 11 wxListBox has a new layout because underlying
NSTableView has a new style with padding so we need to move
the point to be tested to another position.

vadz committed:
Regenerate toolbar sample bitmaps in 32 and 64 pixel sizes

They're more appropriate than 16 and 32 pixels nowadays.

Also add SVG versions of the icons, from which these bitmaps were
created.

oneeyeman1 committed:
Adding the cleanup

oneeyeman1 committed:
Adding the cleanup

joukj committed:
Add wxUSE_SPELLCHECK to setup for OpenVMS

vadz committed:
Merge branch 'filename-misc-fixes'

Fix handling of single letter UNC paths and other wxFileName
improvements.

See https://github.com/wxWidgets/wxWidgets/pull/2519

vadz committed:
Add unit test for wxWindowList::Find()

Check that Find() of a list declared using WX_DECLARE_LIST_3 (as
wxWindowList is) can be passed a pointer to the base class object and
still compiles and works correctly.

vadz committed:
Attempt to work around build problem under OpenVMS

OpenVMS build was broken by 303d899cc0 (Make STL-based wxList more
compatible with the default one, 2021-09-02), try to blindly fix it by
casting the element to the same type it had had before.

a-wi committed:
Invalidate selection of item being removed from wxChoice (wxOSX)

We need to explicitly invalidate selection of item being removed
because it's not done by wxChoiceWidgetsImpl::RemoveItem().

kosh543 committed:
Add wxStyledTextCtrl XRC handler

Closes https://github.com/wxWidgets/wxWidgets/pull/2530

Co-authored-by: VZ <vz-github@zeitlins.org>

a-wi committed:
Execute ChoiceTestCase on wxMac

vadz committed:
Add wxHAS_IMAGE_RESOURCES defined under MSW and Mac

Unlike the existing wxHAS_IMAGES_IN_RESOURCES constant defined only
under MSW, the new one is also defined under Mac and could be defined
for the other platforms/ports later (e.g. wxQt could probably support it
too).

It's unfortunate that two very similar constants are needed, but it
doesn't seem wise to change the meaning of the existing constant, as
this would change how the commonly used wxICON() and wxBITMAP() macros
behave and would break all our own samples that use them for their frame
icon but don't embed this icon into the bundle resources under Mac.

Do change the toolbar sample to use the new constant however, as this
one does include the bitmaps it uses in its bundle under Mac.

a-wi committed:
Don't run CI builds on Ubuntu 16 not supported any longer

Support for Ubuntu 16.04 ended on September 20, 2021.
See: https://github.blog/changelog/2021-04-29-github-actions-ubuntu-16-04-lts-virtual-environment-will-be-removed-on-september-20-2021/

See https://github.com/wxWidgets/wxWidgets/pull/2531

vadz committed:
Fix refresh problems when scrolling with mousewheel in wxUniv/MSW

Use native scrolling under MSW, just as we already did under X11,
instead of our own version which doesn't work correctly when scrolling
using the mouse wheel.

See https://github.com/wxWidgets/wxWidgets/pull/2525

vadz committed:
Merge branch 'list-std'

Fix build in STL mode under OpenVMS.

See https://github.com/wxWidgets/wxWidgets/pull/2528

vadz committed:
Merge branch 'toolbar-high-dpi'

Show how to use bitmaps appropriate for the current DPI in the toolbar
sample using the current API.

See https://github.com/wxWidgets/wxWidgets/pull/2520

vadz committed:
Fix self-assignment bug in wxObjectDataPtr

wxObjectDataPtr::operator=() didn't handle self-assignment correctly,
i.e. could delete the object when doing "x = x".

Fix this by incrementing the reference count before decrementing it on
possibly the same object to ensure that it never reaches 0.

vadz committed:
Replace multiple RescaleBitmap() with wxBitmap::Rescale()

Define a single function and use it in both wxArtProvider and (the
generic implementation of) wxSearchCtrl instead of repeating the same
code elsewhere.

Note that another, but slightly different, version of RescaleBitmap()
still remains in wxPropertyGrid.

Deprecate undocumented wxArtProvider::RescaleBitmap() which is
completely useless now.

No real changes, this is just a refactoring.

vadz committed:
Use high quality in wxBitmap::Rescale()

This function is currently used for relatively small icons in
wxArtProvider and wxSearchCtrl, so it shouldn't be a problem to use
wxIMAGE_QUALITY_HIGH from performance point of view and if it ever does
become a problem, the application could easily use wxImage::Rescale()
with a different option directly instead.

joukj committed:
Add uilocale in OpenVMS makefiles

vadz committed:
Merge branch 'bitmap-rescale'

Add wxBitmap::Rescale() and use it instead of RescaleBitmap().

See https://github.com/wxWidgets/wxWidgets/pull/2534

rlbxku1r committed:
wxMBConv_iconv: Fix wrong buffer size

This is an oversight in commit 878c265.

oneeyeman1 committed:
Adding error logging - all windows functions can fail

oneeyeman1 committed:
Adding error logging - all windows functions can fail

oneeyeman1 committed:
Error handling

oneeyeman1 committed:
Error handling

oneeyeman1 committed:
Fix merge conflict

tbultel committed:
Fix missing mouse events when using gesture events in wxGTK

When touch events were enabled, normal mouse move and click events were
not generated any longer. Fix this by doing the following two things:

First, emulate the LeftDown, LeftUp, and Motion events, that are no
longer triggered when using the "touch-event".

In addition, remove GDK_BUTTON1_MASK in the press callback, that is
never set when using a mouse, but that is set for some reason by GTK
when using a touchscreen.

See https://github.com/wxWidgets/wxWidgets/pull/2539

Closes #19265.

Signed-off-by: Thierry Bultel <tbultel@free.fr>

IhateTrains committed:
Support for VS 2022 in wxwidgets.props

Kvaz1r committed:
Fix handling wxSHOW_SB_ALWAYS in SetScrollbar() in wxUniv

Range may be -1 when wxSHOW_SB_ALWAYS is used, see
wxScrollHelper::DoAdjustScrollbar(), so check for it to avoid spurious
asserts in this case.

Closes https://github.com/wxWidgets/wxWidgets/pull/2540

oneeyeman1 committed:
Adding temporary printfs to find the cause of test failure

oneeyeman1 committed:
Adding temporary printfs to find the cause of test failure

oneeyeman1 committed:
Resolve merge conflict

PBfordev commented:
@oneeyeman1, as I have already commented on this feature A LOT, I was hesitant to review it again, as you could think I am trying to pick on you (I am not and I believe it or not, I do dislike to criticize). In the end, I decided to post my observations, hoping you would not mind and perhaps can find something useful in them.

I am familiar only with MSW, so all my comments are based on MSW implementation, except for one: I am not sure if `std::vector<wxString>` is the new official API replacement of `wxArrayString` or not, so I left it there; however, the downside of using it is that it does not have simple conversion to `wxArrayString`. Also, should not we `#include <vector>` somewhere before #including the platform-specific header?

### The deleted paths retrieving API
1. On Windows, the Recycle Bin can contain the same path multiple times, with differing delete dates, i.e., the content of the deleted file/folder can be different even when the path is the same. The delete date is available via `SCID_DateDeleted` property (see e.g. https://devblogs.microsoft.com/oldnewthing/20110830-00/?p=9773). I believe having the delete date for a path available is important, so the last deleted path is restored as probably expected instead of the version deleted before that. 
2. The method name is `wxDesktopEnv::GetFilesInRecycleBin()`; however, the current code allows also retrieving folders (see `SHCONTF_FOLDERS` flag in the `IShellFolder::EnumObjects()` call). I think the code should support folders and if so, perhaps it should be also able to obtain not only the path (and the deleted date) but also whether the path is a folder of a file?

### The actual MSW implementation
I find the code somehow odd, not sure if you pushed the intended version? AFAICT, the code, among else:
1.	Does not work, for example at the beginning of `wxDesktopEnv::GetFilesInRecycleBin()` you are testing the result of `SHGetKnownFolderIDList()` as if it returned true on success; however, it returns 0 (`S_OK`) when it succeeds.
2.	Contains dead code, such as retrieving the name of the Recycle Bin folder which is then never used.
3.	Uses `SHGetKnownFolderIDList()` which is available only for Vista+, while AFAIK wxWidgets 3.1 unfortunately still has to support Windows XP.
4.	Uses raw pointers instead of RAII wrappers provided by wxWidgets.
5.	Is quite difficult to follow, for example look at the length and scope level depth of `wxDesktopEnv::RestoreFromRecycleBin()`.

Instead of commenting on the perceived implementation issues individually, I attempted to rewrite the code, with focus on addressing the points listed above. You can my attempt here: https://github.com/wxWidgets/wxWidgets/commit/69bd24f05a94079445a4335729a7696b8fec0268 I hope it could be of some help.

By the way, Windows can still display a confirmation dialog when restoring the Recycle Bin item would overwrite an existing path, despite all the flags used to suppress any user interaction. Also, when the user chooses not to restore the path, the API still returns success so the code cannot tell whether the item was actually restored or the user decided not to.

Last but not least, my implementation restores the first found deleted path, which may not be the last deleted and thus not what the user expects, see the API discussion section.

FWIW, here is a simple minimal application to test enumerating the Recycle Bin and restoring an item inside it: https://gist.github.com/PBfordev/7a4865bf9ec7be14e5c716c5d81c5385

### Conclusion
I believe that the current implementation is not optimal and could be improved. But before doing anything, I would first wait for the confirmation that the maintainers are willing to accept a feature for retrieving and restoring deleted paths, as AFAIK this goes beyond the originally accepted function for moving paths to the trash.

oneeyeman1 commented:
@PBfordev ,
> I still do not understand the API, i.e., why have a class with a single, static, method. I guess this is supposed to be extended in the future.

It is possible to have a function here to create a shortct on the Desktop. That new function will belong to this class.
This is just one example.

> Even if so, why not submit it complete so we have a big picture available?

Because my goad is to implement "Recycle Bin" functionality.
Also the old ticket contained some stuff that could be implemented in this class.

> Nevertheless, I added a bunch of (mostly) nitpick comments.

Thank you.
